pandas/core/groupby/groupby.py

```diff
@@ -3290,6 +3304,103 @@ def max(
                 npfunc=np.max,
             )
 
+    def _idxmin_idxmax(
+        self,
+        how: Literal["idxmax", "idxmin"],
+        numeric_only: bool = False,
+        skipna: bool = True,
+        ignore_unobserved: bool = False,
+    ) -> NDFrameT:
+        """Compute idxmax/idxmin.
+
+        Parameters
+        ----------
+        how: {"idxmin", "idxmax"}
+            Whether to compute idxmin or idxmax.
+        numeric_only : bool, default False
+            Include only float, int, boolean columns.
+        skipna : bool, default True
+            Exclude NA/null values. If an entire row/column is NA, the result
+            will be NA.
+        ignore_unobserved : bool, default False
+            When True and an unobserved group is encountered, do not raise. This used
+            for transform where unobserved groups do not play an impact on the result.
+
+        Returns
+        -------
+        Series or DataFrame
+            idxmax or idxmin for the groupby operation.
+        """
+        # unobserved is passed to the Cython code and mutated
+        unobserved = np.ones(self.ngroups, dtype="bool")
+
+        def post_process(res):
```

Comment:
The only other function we have with a post-processor is quantile, which looks like it captures the arguments it needs whereas this does not. Are you expecting to port quantile over to this pattern?

Generally would be better to have the enclosed function capture the arguments it needs to reference as local variables. I might be missing the larger picture to how this is supposed to function

Reply:
> Are you expecting to port quantile over to this pattern?

I have no plans in this regard.

I think you're suggesting to change this to `def post_processing(res, unobserved):`. This would require changing `_cython_agg_general` from

```
if post_process is not None:
    res = post_process(res)
```

to

```
if post_process is not None:
    res = post_process(res, **kwargs)
```

which seems like an odd pattern.

Reply:
I do not know. Do you have an alternative that is more palatable? I'm happy to implement, but I do not see one.

My original implementation was (slightly) more performance and we did not have to do this.

Reply:
Friendly ping @WillAyd - this is blocking some future cleanups of groupby that I'd like to get in.

Reply:
I'll try to take another pass this weekend, but yea at a minimum I think we should pass the arguments being used into the post processing function. Otherwise it's not really clear to the next developer what local variables are being referenced from outside of this function

Reply:
I agree that that is a concern, but I find the practicality of using variables from the defining scope wins out. As long as functions aren't unwieldly long, I've never had issues in practice.

A quick glace of methods that already do this:

 - `GroupBy._cython_agg_general`
 - `GroupBy.count`
 - `GroupBy.quantile`
 - `GroupBy.value_counts`
 - `DataFrameGroupBy._cython_transform`
 - `DataFrame.select_dtypes`
 - `DataFrame._combine_frame`
 - `DataFrame._align_for_op`

This isn't exhaustive, and I skipped over the shorter examples (functions that were 1-3 lines long).

Reply:
I've tried to move us away from the pre/post processors where at all possible bc they can be tough to debug.  Note the only remaining one is i quantile where it isn't really even necessary anymore, could be inlined into blk_func.

What if instead we had a flag that disabled the `out = self._wrap_aggregated_output(res)` line in cython_agg_general?

Reply:
> bc they can be tough to debug

Curious why this is. I haven't run into it personally.

> What if instead we had a flag that disabled the `out = self._wrap_aggregated_output(res)` line in cython_agg_general?

Definitely viable, but then we'd have to repeat the remaining logic of `_cython_agg_general` within `_idxmin_idxmax`. It isn't much, but I personally prefer trying to leave that as consolidated as possible. If it gets updated, then forgetting to do so for idxmin/idxmax introduces an inconsistency. And my preference here is somewhat hinged upon not running into issues with debugging a post_processing method, so I may just not be seeing the full issue there.

Another option I thought worse was adding the post_processing function directly to `WrappedCythonOp`. If we did it there, we could utilize the standard `counts` which is returned from the Cython code. My reasoning for not liking this was trying to keep all of the compute either in the groupby method `_idxmin_idxmax` itself or in the Cython code, rather than have bits be within the intermediate layers (as much as possible). But this would eliminate both the additional mask which I find odd, the use of a variable from the defining scope which @WillAyd doesn't like, and the post-processing which you don't like. So it's making me think that's now a much better option.

Reply:
What if we instead had a method `_wrap_idxmax_idxmin`? This would get rid of post-processing (sort of) and we could pass the mask directly. It would be called from `_cython_agg_general` where post-processing is now, branching on `how`.